aiida/engine/processes/workchains/workchain.py

```diff
@@ -124,16 +124,45 @@ def insert_awaitable(self, awaitable):
         :type awaitable: :class:`aiida.engine.processes.workchains.awaitable.Awaitable`
         """
         self._awaitables.append(awaitable)
+
+        # Already assign the awaitable itself to the location in the context container where it is supposed to end up
+        # once it is resolved. This is especially important for the `APPEND` action, since it needs to maintain the
+        # order, but the awaitables will not necessarily be resolved in the order in which they are added. By using the
+        # awaitable as a placeholder, in the `resolve_awaitable`, it can be found and replaced by the resolved value.
+        if awaitable.action == AwaitableAction.ASSIGN:
+            self.ctx[awaitable.key] = awaitable
+        elif awaitable.action == AwaitableAction.APPEND:
+            self.ctx.setdefault(awaitable.key, []).append(awaitable)
```

Comment:
Might be worth having a quick check in here just to make sure it hasn't been appended before.  Just something like
```
assert awaitable.pk not in (await.pk for await in awaitables)
```

Reply:
Why? I don't see the use case, but if a user _wants_ to add the same subprocess twice, why not let them? When asserting, we are mostly doing this for internal consistency, no? Here we would be basically checking whatever the user returned from a step, because the `insert_awaitable` is only called when `to_context` or `ToContext` is used with a child process

Reply:
Yup, agreed they might want to be able to do it but this PR doesn't support it.   Unless I've misunderstood, when resolving it looks like you interate the placeholders container from start to end and so will find the _frist_ matching entry which will be the corresponding node the second time around and just be overwritten leaving the second occurrence unresolved.

Agreed re: assert, this should really be some kind of timetime error, just didn't want to push an additional line of code change on you.

Reply:
ah damn you're right. Well that is just a straight up bug! When checking the placeholders I should also check that it actually is still an awaitable and not a node, in addition to checking the pk, right?

Reply:
Yeah, I was thinking of a type check as well...